### ISSUE TITLE
Effects are visible through the water surface

### ISSUE THREAD

**Issue by Einheit-101** [multitheftauto/mtasa-blue#698]
**Describe the bug**
This old rendering bug is annoying for a long time now. Effects ignore the water in front of them.

**To reproduce**
1. Go to editor 
2. Go into water
3. crun x, y, z = getElementPosition(me) createEffect("prt_bubble", x, y, z-1, 0, 0, 0, 2000)
4. See bubbles being drawn in front of water

**Expected behaviour**
Effects should render correctly

**Screenshots**
![image](https://user-images.githubusercontent.com/923242/50401633-14f41d00-0788-11e9-9b98-6f4c6d8da960.png)

**Multi Theft Auto v1.5.6-release-15932 (Windows)**

**Additional context**
Is there maybe a shader that can be applied to effects textures that fixes this issue?

**Comment by Wolfee-J** [comment]
This is due to the rending order.

**Comment by AlexTMjugador** [comment]
If what CodyJL said is right, maybe using [setWaterDrawnLast](https://wiki.multitheftauto.com/wiki/SetWaterDrawnLast) makes a difference?

**Comment by Einheit-101** [comment]
No! Effects are still in front of water with setWaterDrawnLast(true)

**Comment by Einheit-101** [comment]
Creating a simple shader and adding this code into the main pass does actually solve this issue:
ZEnable = true;
ZWriteEnable = true;
ZFunc = 4;

...however since water is now writing to depth buffer, it causes zfighting with anything that is close and planar to it (including boat wake and other water effects).

![image](https://user-images.githubusercontent.com/923242/49098757-8911d280-f267-11e8-8b77-ec8992294637.png)


I think the final solution would be applying another shader to these water effects (duh...) that gives them so much depth bias that they will draw on top of the water surface...

**Comment by Wolfee-J** [comment]
Can you check if that shader fixes the bug of not being able to see water through car windows @Einheit-101

**Comment by Einheit-101** [comment]
Sure. It doesnt fix the issue. 
Default GTA Water with simple zWriteEnable shader: 

![image](https://user-images.githubusercontent.com/923242/49098778-98911b80-f267-11e8-83ce-2c6dfa6189c3.png)

Water still not visible through car window, but water surface can be illuminated by dynamic lighting

Complex water shader with zWriteEnable = same.

![image](https://user-images.githubusercontent.com/923242/49098787-9fb82980-f267-11e8-82d1-c1139433e96b.png)

Water doesnt just refuse to render through car windows, actually it doesnt render through vehicles at all.
![image](https://user-images.githubusercontent.com/923242/49098797-a3e44700-f267-11e8-8178-98d0bd5bd4f0.png)

**Comment by Wolfee-J** [comment]
Alright, reconfirms my belief that SAs rendering order is crap.

**Comment by Einheit-101** [comment]
setWaterDrawnLast on top of this shader doesnt work either. It looks like the water is simply culled away. Yes, GTA render order is messed up.
